git-utils/src/lib.rs

```diff
@@ -0,0 +1,160 @@
+use std::fs::File;
+use std::io::prelude::*;
+use std::path::{Path, PathBuf};
+
+use git2::{Branches, BranchType, Commit, Repository, Signature};
+use git2::build::RepoBuilder;
+use tempdir::TempDir;
+
+pub struct TestGit {
+    pub dir: PathBuf,
+    pub remote: Repository,
+    pub local: Repository,
+}
+
+// Initialise a git repo with a remote in the state that
+// remote - 3 commits, initial, first commit, second commit
+// local - 2 commits, initial, first commit
+// So that if we fetch on local then it should know about the second file
+pub fn initialise_git_repo(path: Option<&PathBuf>) -> TestGit {
+    let tmp_dir = TempDir::new("repo").expect("Temp dir created").into_path();
+    let remote_path = tmp_dir.join("remote");
+    let local_path = tmp_dir.join("local");
+    match path {
+        Some(p) => copy_recursively(p, &remote_path),
+        None => std::fs::create_dir(&remote_path),
+    }.unwrap();
+    std::fs::create_dir(&local_path).unwrap();
+
+    let remote = Repository::init(&remote_path).unwrap();
+    create_initial_commit(&remote);
+    create_file(&remote_path, "new_file");
+    git_add_all(&remote);
+    git_commit(&remote, "First commit");
+
+    let local = git_clone_local(&remote, &local_path);
+
+    create_file(&remote_path, "new_file2");
+    git_add_all(&remote);
+    git_commit(&remote, "Second commit");
+
+    let default_branch = git_branch(&remote);
+    git_checkout(&remote, "other", true);
+    create_file(&remote_path, "new_file3");
+    git_add_all(&remote);
+    git_commit(&remote, "Third commit");
+    git_checkout(&remote, &default_branch, false);
```

Comment:
As an alternative to working with the worktree and the index and having to switch branches around, you could also use the lower level APIs that just modify the contents of `.git` directly, eg. using https://docs.rs/git2/latest/git2/build/struct.TreeUpdateBuilder.html

Either options are fine, the lower level APIs are just a bit less stateful

Reply:
Ok had a quick look at that but a little confused about low level stuff. Tempted to stick with this since it is working but we can move to lower level if it proves brittle at all in the future?

Reply:
Yeah that’s fine.